vizro-ai/src/vizro_ai/plot/_response_models.py

```diff
@@ -0,0 +1,233 @@
+
+try:
+    from pydantic.v1 import BaseModel, Field, PrivateAttr, create_model, validator
+except ImportError:  # pragma: no cov
+    from pydantic import BaseModel, Field, PrivateAttr, create_model, validator
+import logging
+import subprocess
+from typing import List, Optional
+
+import pandas as pd
+import plotly.express as px
+import plotly.graph_objects as go
+
+from vizro_ai.plot._utils._safeguard import _safeguard_check
+
+ADDITIONAL_IMPORTS = [
+    "import vizro.plotly.express as px",
+    "import plotly.graph_objects as go",
+    "import pandas as pd",
+    "import numpy as np",
+    "from vizro.models.types import capture",
+]
+CUSTOM_CHART_NAME = "custom_chart"
+
+
+def _format_and_lint(code_string: str) -> str:
+    # Tracking https://github.com/astral-sh/ruff/issues/659 for proper python API
+    # Good example: https://github.com/astral-sh/ruff/issues/8401#issuecomment-1788806462
+    linted = subprocess.check_output(
+        ["ruff", "check", "--fix", "--exit-zero", "--silent", "--isolated", "-"], input=code_string, encoding="utf-8"
+    )
+    formatted = subprocess.check_output(
+        ["ruff", "format", "--silent", "--isolated", "-"], input=linted, encoding="utf-8"
+    )
+    return formatted
+
+
+def _exec_code(code: str):
+    """Execute code and return the local dictionary."""
+    # globals needed to access imports, they will not be modified by exec
+    # ldict is used to store the chart function
+    # potentially possible to restrict globals to only needed imports, but that is tricky
+    ldict = {}
+    exec(code, globals(), ldict)  # nosec
+    return ldict
+
+
+class ChartPlanStatic(BaseModel):
+    """Chart plan model."""
+
+    chart_type: str = Field(
+        ...,
+        description="""
+        Describes the chart type that best reflects the user request.
+        """,
+    )
+    imports: List[str] = Field(
+        ...,
+        description="""
+        List of import statements required to render the chart defined by the `chart_code` field.
+        """,
+    )
+    chart_code: str = Field(
+        ...,
+        description=f"""
+        Python code that generates a generates a plotly go.Figure object. It must fulfill the following criteria:
+        1. Must be wrapped in a function named `{CUSTOM_CHART_NAME}`
+        2. Must accept a single argument `data_frame` which is a pandas DataFrame
+        3. Must return a plotly go.Figure object
+        4. All data used in the chart must be derived from the data_frame argument, all data manipulations
+        must be done within the function.
+        """,
+    )
+    chart_insights: str = Field(
+        ...,
+        description="""
+        Insights to what the chart explains or tries to show. Ideally concise and between 30 and 60 words.""",
+    )
+    code_explanation: str = Field(
+        ...,
+        description="""
+        Explanation of the code steps used for `chart_code` field. Ideally a concise bulleted step-by-step list.""",
```

Comment:
```suggestion
        Explanation of the code steps used for `chart_code` field.""",
```
Seems solved the incorrect response type